### PR TITLE
Document 3D Abstraction Layer

### PR DESCRIPTION
**Pull request by Fexty12573**
Documents the 2 levels of 3D abstraction layers in the game. Below a "quick" write-up about the system.

TL;DR: Easy 3D model loading and drawing API.

## Abstraction Layers
The lowest level 3D is interacting directly with the hardware via NitroSDK. NitroSystem provides a slight abstraction over this with `G3D`.

The game has 2 additional layers of abstraction over this:
### 1 - Easy3D
The first layer of abstraction is the Easy3D System (`easy3d.h`). Which provides basic functionality for:
- Loading Models and Textures
- Binding textures to models
- Drawing models with any transform
- Setting up and releasing a 3D graphics state

To elaborate on the last point: Before using any 3D graphics, a "3D Graphics State" must be initialized. The struct for this is currently called `GenericPointerData`. The Easy3D system provides an easy method to set up a simple gfx state via `Easy3D_Init` (With `Easy3D_Shutdown` as its counterpart), which works for most scenarios.

### 2 - Easy3DObject
The second layer of abstraction is `Easy3DObject` (`easy3d_object.h`). This API provides a streamlined interface for:
- Loading models, textures, and animations
- Binding a texture to a model
- Binding one or more animations to a model (both bone animations and texture animations)
- Updating the objects transform (the object itself keeps track of that)
- Updating animation state
- Drawing objects
- Cleanup

One thing to keep in mind is that this system by itself does not establish a 3D GFX State.
This system is also not used everywhere. In a lot of places Easy3D is used directly instead of this object oriented interface.

## Example
The following is an example for using the `Easy3DObject` API for:
1. Loading a model with textures, animation, and texture animation
2. Binding them all together
3. Updating the animations
4. Drawing the model

I will be using the Giratina model from the title screen for this.

### Loading the data
The first thing needed is some static storage to hold our data:
```c
Easy3DObject giratinaObj;
Easy3DModel giratinaModel;
Easy3DAnim giratinaAnim; // Model Animation
Easy3DAnim giratinaTexAnim; // Texture Animation
NNSFndAllocator allocator; // Needed for Animations
```
Next we load all of the data:
```c
// Open the title NARC
NARC *narc = NARC_ctor(NARC_INDEX_DEMO__TITLE__TITLEDEMO, HEAP_ID_FIELD);

// Load the model from the title screen NARC. Member index 1 is the model data.
// There is also Easy3DModel_Load which takes a NARC index and a member index.
Easy3DModel_LoadFrom(&giratinaModel, narc, 1, HEAP_ID_FIELD);
Easy3DObject_Init(&giratinaObj, &giratinaModel);

// Initialize the Allocator used by the animations
Heap_FndInitAllocatorForExpHeap(&allocator, HEAP_ID_FIELD, 4);

// Load the model animation with member index 2.
Easy3DAnim_LoadFrom(&giratinaModelAnim, &giratinaModel, narc, 2, HEAP_ID_FIELD, &allocator);
// Bind the animation to the object
Easy3DObject_AddAnim(&giratinaObj, &giratinaModelAnim);

// Do the same for the texture animation
Easy3DAnim_LoadFrom(&giratinaTexAnim, &giratinaModel, narc, 0, HEAP_ID_FIELD, &allocator);
Easy3DObject_AddAnim(&giratinaObj, &giratinaTexAnim);

NARC_dtor(narc);
```
This is all that needs to be done in terms of setup, now the model is ready to draw (provided a 3D GFX State has been set up).

### Drawing the Model
Before drawing there's a few things that should be configured on the object, one of them being the position of the model obviously. For simplicity's sake I will just set the models position to the player's position.
```c
const VecFx32 *pos = PlayerAvatar_PosVector(fieldSystem->playerAvatar);
Easy3DObject_SetPosition(&giratinaObj, pos->x, pos->y, pos->z);

// Make sure the model actually gets rendered
Easy3DObject_SetVisibility(&giratinaObj, TRUE); 

// The model is pretty big so scale it to half its size
Easy3DObject_SetScale(&giratinaObj, FX32_CONST(0.5), FX32_CONST(0.5), FX32_CONST(0.5)); 
```
Now we can actually render the model:
```c
// Update the animations
// Here we advance the animation by one frame.
// There is also Easy3DAnim_Update which does not loop the animation
Easy3DAnim_UpdateLooped(&giratinaModelAnim, FX32_ONE);
Easy3DAnim_UpdateLooped(&giratinaTexAnim, FX32_ONE);

// Draw the model
Easy3DObject_Draw(&giratinaObj);
```
All of that results in the following:

https://github.com/pret/pokeplatinum/assets/60443001/ba162c62-e64a-4cd6-850f-414774f19bfd

Obviously the rotation and resizing is not outlined above. The code for this is the following:
```c
u16 angle = Easy3DObject_GetRotation(&giratinaObj, ROTATION_AXIS_Y);
angle = (angle + 100) % 0xFFFF;
Easy3DObject_SetRotation(&giratinaObj, angle, ROTATION_AXIS_Y);

// Resize using L and R
if (gCoreSys.heldKeys & PAD_BUTTON_R) {
    VecFx32 scale;
    Easy3DObject_GetScale(&giratinaObj, &scale.x, &scale.y, &scale.z);
    scale.x = FX_Mul(scale.x, FX32_CONST(1.01));
    scale.y = FX_Mul(scale.y, FX32_CONST(1.01));
    scale.z = FX_Mul(scale.z, FX32_CONST(1.01));
    Easy3DObject_SetScale(&giratinaObj, scale.x, scale.y, scale.z);
} else if (gCoreSys.heldKeys & PAD_BUTTON_L) {
    VecFx32 scale;
    Easy3DObject_GetScale(&giratinaObj, &scale.x, &scale.y, &scale.z);
    scale.x = FX_Mul(scale.x, FX32_CONST(0.99));
    scale.y = FX_Mul(scale.y, FX32_CONST(0.99));
    scale.z = FX_Mul(scale.z, FX32_CONST(0.99));
    Easy3DObject_SetScale(&giratinaObj, scale.x, scale.y, scale.z);
}
```

### CODE REVIEW

File: docs/3d_rendering.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
+## 3D Rendering
```
Comment (lhearachel) [inline docs/3d_rendering.md:1]: **chore:** Make sure to add a link to this to `docs/index.md`, similar to the `bugs_and_glitches` doc. I would create a new subsection in the `index` to house documents like this, e.g. `Subsystems`

File: include/easy3d.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#ifndef POKEPLATINUM_EASY3D_H
+#define POKEPLATINUM_EASY3D_H
+
+#include <nitro/fx/fx.h>
+#include <nnsys.h>
+
+void Easy3D_LoadModelFromPath(const u8 heapID, const char *path, NNSG3dResMdl **outModel, NNSG3dResFileHeader **outResource);
+void Easy3D_InitRenderObjFromPath(const u8 heapID, const char *path, NNSG3dRenderObj *obj, NNSG3dResMdl **outModel, NNSG3dResFileHeader **outResource);
+void Easy3D_InitRenderObjFromResource(NNSG3dRenderObj *renderObj, NNSG3dResMdl **model, NNSG3dResFileHeader **resource);
+BOOL Easy3D_IsTextureUploadedToVRam(NNSG3dResTex *texture);
```
Comment (lhearachel) [inline include/easy3d.h:10]: **polish:** I would capitalize `VRAM`

File: src/easy3d.c [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+#include <nnsys.h>
+
+#include "struct_defs/struct_0207C690.h"
+
+#include "unk_02017728.h"
+#include "easy3d.h"
+#include "gx_layers.h"
+#include "unk_02024220.h"
+
+void Easy3D_LoadModelFromResource(NNSG3dResMdl **outModel, NNSG3dResFileHeader **resource);
```
Comment (lhearachel) [inline src/easy3d.c:12]: **question:** Is this function not static? If it's not, then move its declaration to the header.

File: src/easy3d.c [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+static void Easy3D_EngineSetup(void);
```
Comment (lhearachel) [inline src/easy3d.c:13]: **polish:** `Easy3D_SetupEngine`

File: src/easy3d.c [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    GXLayers_EngineAToggleLayers(GX_PLANEMASK_BG0, TRUE);
+
+    G2_SetBG0Priority(1);
+    G3X_SetShading(GX_SHADING_TOON);
+    G3X_AntiAlias(TRUE);
+    G3X_AlphaTest(FALSE, 0);
+    G3X_AlphaBlend(TRUE);
+    G3X_EdgeMarking(FALSE);
+    G3X_SetFog(FALSE, GX_FOGBLEND_COLOR_ALPHA, GX_FOGSLOPE_0x8000, 0);
+    G3X_SetClearColor(GX_RGB(0, 0, 0), 0, 0x7fff, 63, FALSE);
```
Comment (lhearachel) [inline src/easy3d.c:97]: **question:** Would it be useful to mark `0x7FFF` as the maximum value for a RGB15 palette?

File: src/easy3d.c [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+{
+    if (texture == NULL) {
+        return FALSE;
+    }
+
+    texture->texInfo.vramKey = 0;
+
+    u32 texRequiredSize, 
+        tex4x4RequiredSize, 
+        paletteRequiredSize;
```
Comment (lhearachel) [inline src/easy3d.c:129]: **polish:** Declare these in the same line.

File: src/easy3d.c [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    BOOL texOk = TRUE;
+    BOOL tex4x4Ok = TRUE;
+    BOOL paletteOk = TRUE;
+    NNSG3dTexKey texKey;
+    NNSG3dTexKey tex4x4Key;
+    NNSG3dPlttKey paletteKey;
+
+    if (texture) {
```
Comment (lhearachel) [inline src/easy3d.c:137]: **question:** Is this `if` necessary? Does the compiler strip it due to the preceding sentinel?

**Review by lhearachel — APPROVED** [review]
Nice work overall. Just a few minor quibbles and questions.

File: src/easy3d.c (hunk 5 above)
Comment (Fexty12573) [reply src/easy3d.c:97]: The 0x7FFF he refers to a depth value, not a color. Point still stands though, a constant would be appropriate. Not sure what to call it though. As far as I know there isn't any constant defined for this in the SDK. `G3X_MAX_DEPTH` maybe or `G3X_DEPTH_MAX`?

File: src/easy3d.c (hunk 7 above)
Comment (Fexty12573) [reply src/easy3d.c:137]: Yeah it doesn't match without it

File: src/easy3d.c (hunk 5 above)
Comment (lhearachel) [reply src/easy3d.c:96]: I like `G3X_DEPTH_MAX`.
